.github/workflows/CI.yml

```diff
@@ -3,66 +3,57 @@ name: CI
-          - {target: ppc,     toolchain: gcc-powerpc-linux-gnu,     cc: powerpc-linux-gnu-gcc,      qemu: qemu-ppc-static     }
-          - {target: ppc64,   toolchain: gcc-powerpc64-linux-gnu,   cc: powerpc64-linux-gnu-gcc,    qemu: qemu-ppc64-static   }
-          - {target: ppc64le, toolchain: gcc-powerpc64le-linux-gnu, cc: powerpc64le-linux-gnu-gcc,  qemu: qemu-ppc64le-static }
-          - {target: s390x,   toolchain: gcc-s390x-linux-gnu,       cc: s390x-linux-gnu-gcc,        qemu: qemu-s390x-static   }
-          - {target: mips,    toolchain: gcc-mips-linux-gnu,          cc: mips-linux-gnu-gcc,         qemu: qemu-mips-static     }
-          - {target: mips64,  toolchain: gcc-mips64-linux-gnuabi64,   cc: mips64-linux-gnuabi64-gcc,  qemu: qemu-mips64-static   }
-          - {target: mipsel,  toolchain: gcc-mipsel-linux-gnu,        cc: mipsel-linux-gnu-gcc,       qemu: qemu-mipsel-static   }
-          - {target: mips64el,toolchain: gcc-mips64el-linux-gnuabi64, cc: mips64el-linux-gnuabi64-gcc,qemu: qemu-mips64el-static }
-          - {target: alpha,   toolchain: gcc-alpha-linux-gnu,       cc: alpha-linux-gnu-gcc,        qemu: qemu-alpha-static   }
-          - {target: arm (u64 slots),     toolchain: gcc-arm-linux-gnueabi,     cc: arm-linux-gnueabi-gcc,      qemu: qemu-arm-static}
-          - {target: aarch64 (u64 slots), toolchain: gcc-aarch64-linux-gnu,     cc: aarch64-linux-gnu-gcc,      qemu: qemu-aarch64-static}
-          - {target: ppc (u64 slots),     toolchain: gcc-powerpc-linux-gnu,     cc: powerpc-linux-gnu-gcc,      qemu: qemu-ppc-static}
-          - {target: ppc64 (u64 slots),   toolchain: gcc-powerpc64-linux-gnu,   cc: powerpc64-linux-gnu-gcc,    qemu: qemu-ppc64-static}
+          - {arch: armhf,   triple: arm-linux-gnueabihf,   qemu: arm     }
+            # - {arch: arm64,   triple: aarch64-linux-gnu,      qemu: aarch64 }
+            #          - {arch: riscv64, triple: riscv64-linux-gnu,      qemu: riscv64 }
+            #          - {arch: powerpc, triple: powerpc-linux-gnu,      qemu: ppc     }
+            #          - {arch: ppc64,   triple: powerpc64-linux-gnu,    qemu: ppc64   }
+          - {arch: ppc64el, triple: powerpc64le-linux-gnu,  qemu: ppc64le }
+            #          - {arch: s390x,   triple: s390x-linux-gnu,        qemu: s390x   }
 
     steps:
       - uses: actions/checkout@v2
       - name: Install QEMU
         # this ensure install latest qemu on ubuntu, apt get version is old
         env:
           QEMU_SRC: "http://archive.ubuntu.com/ubuntu/pool/universe/q/qemu"
-          QEMU_VER: "qemu-user-static_4\\.2-.*_amd64.deb$"
+          QEMU_VER: "qemu-user-static_5\\.0-.*_amd64\\.deb$"
         run: |
           DEB=`curl -s $QEMU_SRC/ | grep -o -E 'href="([^"#]+)"' | cut -d'"' -f2 | grep $QEMU_VER | tail -1`
           wget $QEMU_SRC/$DEB
           sudo dpkg -i $DEB
+      - name: Download Libc support libraries
+        env:
+          LIBC_SRC: "http://ports.ubuntu.com/pool/main/g/glibc"
```

Comment:
We could also get a couple more architectures by using https://packages.debian.org/buster/libc6 instead; or get the complete toolchain from https://toolchains.bootlin.com (just need `/lib`), or by building it ourselves, and saving the useful bits to the cloud.

```
wget https://ftp.gnu.org/gnu/glibc/glibc-2.31.tar.xz
tar xf glibc-2.31
mkdir glibc-2.31/build
cd glibc-2.31/build
CFLAGS='-g0 -O2'  ../configure --build=`../scripts/config.guess` --host=powerpc64le-linux-gnu --prefix=/etc/qemu-binfmt/ppc64le
make -j
make install DESTDIR=~+/staging
cd staging/etc/qemu-binfmt/ppc64le
rm -r lib/*.a lib/*.o lib/gconv # delete some things we don't use
tar cjf libc6-2.31-ppc64le.tar.bz2 lib
```
```
tar xf libc6-2.31-ppc64le.tar.bz2
sudo mkdir -p /etc/qemu-binfmt/ppc64le
sudo mv lib /etc/qemu-binfmt/ppc64le
sudo ln -s lib /etc/qemu-binfmt/ppc64le/lib64
```

Reply:
Any thought on this migrate to docker ?

Reply:
I don't know much about docker. It could work, or it could also be overkill. I tried a simple chroot at one point, and quickly realized it'd be aggravating to configure since it messes up all of the absolute build paths and needs a full OS installed, vs. just needing glibc installed.

Reply:
I give docker a try, looks like it get can preinstall all the toolchain.

```docker
FROM ubuntu:20.04
RUN apt update && \
    apt install -y \
        build-essential \
        ninja-build \
        curl \
        wget \
        git

ENV QEMU_SRC="http://archive.ubuntu.com/ubuntu/pool/universe/q/qemu"
ENV QEMU_VER="qemu-user-static_5\\.0-.*_amd64\\.deb$"
RUN DEB=`curl -s $QEMU_SRC/ | grep -o -E 'href="([^"#]+)"' | cut -d'"' -f2 | grep $QEMU_VER | tail -1` && \
          wget $QEMU_SRC/$DEB && \
          dpkg -i $DEB

RUN apt install gcc-arm-linux-gnueabi -y
```

Reply:
Stuck with github action not supporting non-root user in container, libuv requires non-root to run test.